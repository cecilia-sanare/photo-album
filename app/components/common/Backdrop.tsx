import classNames from 'classnames';
import * as styles from './Backdrop.module.scss';
import { useCachedState } from '../../hooks/use-cached-state';

type BackdropProps = {
  opened?: boolean;
  onClose?: () => void;
}

export function Backdrop({
  opened = false,
  onClose
}: BackdropProps) {
  const [internallyOpened, setInternallyOpened] = useCachedState<boolean>(() => opened, [opened]);

  return (
    <div
      data-testid='backdrop'
      className={classNames(
        styles.backdrop,
        internallyOpened && styles.opened
      )}
      onClick={() => {
        setInternallyOpened(false);
        onClose?.();
      }}
    />
  )
}
