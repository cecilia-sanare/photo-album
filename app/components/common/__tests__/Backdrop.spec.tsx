import { fireEvent, render } from '@testing-library/react';
import { Backdrop } from '../Backdrop';

describe('<Backdrop />', () => {
  it('should render the backdrop', () => {
    const component = render(<Backdrop />);

    expect(component.baseElement).toMatchSnapshot();
  });

  it('should close when the backdrop is clicked', async () => {
    const onClose = jest.fn();
    const component = render(<Backdrop opened onClose={onClose} />);

    expect(onClose).toHaveBeenCalledTimes(0);

    await fireEvent.click(component.getByTestId('backdrop'));

    expect(component.getByTestId('backdrop')).not.toHaveClass('opened');
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  describe('prop(opened)', () => {
    it('should open the backdrop', () => {
      const component = render(<Backdrop opened />);

      expect(component.getByTestId('backdrop')).toHaveClass('opened');
    });

    it('should close the backdrop', () => {
      const component = render(<Backdrop />);

      expect(component.getByTestId('backdrop')).not.toHaveClass('opened');
    });
  });
});
