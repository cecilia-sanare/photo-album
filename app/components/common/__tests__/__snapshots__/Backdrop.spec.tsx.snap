// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Backdrop /> should render the backdrop 1`] = `
<body>
  <div>
    <div
      class="backdrop"
      data-testid="backdrop"
    />
  </div>
</body>
`;
